### PR TITLE
Consume changes go-mod-core-contracts v0.1.56

### LINKED ISSUE #2439: One attempt to decode 9 bytes of malformed CBOR data can exhaust memory
# 🐞 Bug Report

### Affected Services

 The issue is located in: 

<p>

edgex-go/internal/core/data/io.go

```
// Read reads and converts the request's CBOR event data into an Event struct
func (cr cborReader) Read(reader io.Reader, ctx *context.Context) (models.Event, error) {
	c := context.WithValue(*ctx, clients.ContentType, clients.ContentTypeCBOR)
	event := models.Event{}
	bytes, err := ioutil.ReadAll(reader)
        ...
	x := codec.CborHandle{}
	err = codec.NewDecoderBytes(bytes, &x).Decode(&event)
```

There are two potential problems related to memory exhaustion:
1.  Small data -- the CBOR library used by edgex-go can exhaust memory in 1 decode attempt of 9 bytes of malformed data.
2.  Large data -- use of ioutil.ReadAll(reader) into memory is risky and not necessary with most CBOR libraries.  At a minimum, io.LimitReader should be used with a reasonable limit appropriate for edgex-go.  Better yet, io.LimitReader can be passed to CBOR libraries for decoding.

### Is this a regression?

No

### Description and Minimal Reproduction

RFC 7049 Section 8 (CBOR Security Considerations) warns that malformed data can be used to exhaust system resources.

In September 2019, oasislabs/oasis-core found tiny CBOR data could exhaust memory and the problem was traced to the same CBOR library being used by edgex-go today (ugorji/go). They fixed it by switching CBOR libraries.

In February 2020, smartcontractkit/chainlink reported an issue on pivitoltracker (I don't have login access) and fixed it in [a GitHub PR titled "Switch to more secure CBOR library"](https://github.com/smartcontractkit/chainlink/pull/2383).

To reproduce, decode 9-byte or 10-byte malformed CBOR data described in Section 8 of RFC 7049.  

Examples of malformed CBOR data to achieve this can be found on GitHub since September 2019 (or possibly earlier if you look beyond Go projects).

## 🔥 Exception or Error

alt text

For info about CBOR and security, see [Section 8 of RFC 7049 (Security Considerations)](https://tools.ietf.org/html/rfc7049#section-8).

For more comparisons, see [fxamacker/cbor](https://github.com/fxamacker/cbor/blob/master/README.md).

### LINKED ISSUE #2480: Integrate changes to system management agent client in go-mod-core-contracts
# 🚀 Feature Request

### Description
 A clear and concise description of the problem or missing capability...
After this work is complete: https://github.com/edgexfoundry/go-mod-core-contracts/issues/222 update edgex-go to consume the new agent/client.

### PR DESCRIPTION
**Pull request by jdharms**
Updates to edgex-go to consume changes to go-mod-core-contracts
from the following PRs:

https://github.com/edgexfoundry/go-mod-core-contracts/pull/233

https://github.com/edgexfoundry/go-mod-core-contracts/pull/231

This is being clubbed into one PR on the edgex-go side because both
PRs on go-mod-core-contracts have already been merged.

Fixes #2439 and Fixes #2480 in edgex-go.

Replaces ugorji/go with fxamacker/cbor for security reasons.

Reflects the system management client functionality being
moved from the "general" client in go-mod-core-contracts
to an "agent" client.

Signed-off-by: Daniel Harms <jdharms@gmail.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

Issue Number:

#2439 #2480

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by x448** [comment]
@jdharms in the old code you didn't modify, the call to `ioutil.ReadAll(reader)` in io.go can lead to denial of service attacks using large payloads (if another program is providing the data, etc.)

```
line 76:	bytes, err := ioutil.ReadAll(reader)
```

It would be safer to use io.LimitReader with a reasonable size limit appropriate for edgex-go and then pass the limited reader directly to the CBOR library for decoding.

Even without io.LimitReader, it's still safer to pass io.Reader directly to the decoder.
```
// optionally use io.LimitReader r with size limit appropriate to edgex-go
... 
decoder := cbor.NewDecoder(reader)  // decoder will read in 512 byte chunks
err := decoder.Decode(&event)
```

**Comment by x448** [comment]
> Note: github.com/fxamacker/cbor has only two maintainers.

@bnevis-i  by comparison, the CBOR library used by edgex-go has zero contributors based on commit activity since August 15, 2019.  For context, security problems were found in Sep 2019.

If small team size (without context) is an indicator of risk, take a look at a 27 contributor CBOR library written in Rust [looking for maintainers](https://github.com/pyfisch/cbor/issues/179) in part because the maintainer says,

> Within the past year it became increasingly clear to me that I don't understand all the details and interactions of different features. This is in part because I accepted pull requests implementing new features but did not review the code thoroughly enough.

### Number of contributors

Contributor counts (based on commits) from Aug 15, 2019 - Apr 8, 2020 are:
* 2 contributors -- fxamacker/cbor (proposed new CBOR library)
* 0 contributors -- ugorji/go (current CBOR library used by edgex-go)

Graphs showing activity from Aug 11, 2019 - Apr 8, 2020:

For fxamacker/cbor:

image

For ugorji/go:

image

Source: 
* https://github.com/fxamacker/cbor/graphs/contributors?from=2019-08-11&to=2020-04-08&type=c
* https://github.com/ugorji/go/graphs/contributors?from=2019-08-11&to=2020-04-08&type=c

File: internal/system/agent/clients/agent.go [hunk 1]
```diff
@@ -17,39 +17,39 @@ package clients
 import (
 	"sync"
 
-	"github.com/edgexfoundry/go-mod-core-contracts/clients/general"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/agent"
 )
 
-// GeneralType is an alias for general.GeneralClient and hides implementation detail.
-type GeneralType general.GeneralClient
+// AgentType is an alias for general.GeneralClient and hides implementation detail.
```
Comment (mkbhanda) [inline internal/system/agent/clients/agent.go:23]: Does it make sense to:
s/general.GeneralClient/agent.AgentClient

**Comment by mkbhanda** [comment]
I agree with both comments @x448 has made. Please use a limited reader. Also consider making the limit a constant that we could possibly configure, defaulting to something we think sensible today. If you like, could make this a separate issue. For the future, do keep patches separate for different issues, easier on the reviewers.

File: internal/system/agent/clients/agent.go (hunk 1 above)
Comment (jdharms) [reply internal/system/agent/clients/agent.go:23]: Yes, thank you.

**Comment by jdharms** [comment]
I will make the change to use a limit reader and re-request.

I clubbed the PRs because currently both changes on the go-mod-core-contracts side have been merged, and they include interface changes (the sma client functionality).  This means any other go-mod-core-contracts changes won't be able to be used by EdgeX unless it incorporates these changes, so I wanted to unblock development there as quickly as possible.

In the future I'll keep them separate.

**Comment by tsconn23** [comment]
DIscussed the above bug with @jdharms and @michaelestrin. The requirements were clarified and Daniel will rework his changes in `go-mod-core-contracts` before creating another integration PR.

**Pull request closed without merging**